compiler/rustc_abi/src/lib.rs

```diff
@@ -1592,24 +1592,33 @@ pub enum TagEncoding<VariantIdx: Idx> {
     /// (so converting the tag to the discriminant can require sign extension).
     Direct,
 
-    /// Niche (values invalid for a type) encoding the discriminant:
-    /// Discriminant and variant index coincide.
+    /// Niche (values invalid for a type) encoding the discriminant.
+    /// Note that for this encoding, the discriminant and variant index of each variant coincide!
+    /// (This gets checked, for example, in [codegen_ssa](https://github.com/rust-lang/rust/blob/df32e15c56f582eb2bdde07711af6271f2ae660b/compiler/rustc_codegen_ssa/src/mir/operand.rs#L485).)
```

Comment:
That assertion only checks that *one* variant (the "untagged" one, whose niche is used) has matching discriminant and variant index.

Reply:
That sounds like we should add this check to the canonical place for codified layout checks, which is here:

https://github.com/rust-lang/rust/blob/102cc2f8835c2ca73ca9a20172e3c6dda0e4d139/compiler/rustc_ty_utils/src/layout/invariant.rs#L8

Reply:
I've tried moving the check into `layout_sanity_check` and it seems to work well. Do you think this could have influence on performance? If so it might belong into its own PR.

Reply:
I assume you moved it into the part of `layout_sanity_check` that is only run when debug assertions is enabled?

Reply:
Ah, I didn't notice that there was such a check. But yes, the new check is only run when the debug assertions are enabled.